doschema/jsvalidation.py

```diff
@@ -0,0 +1,229 @@
+        """Process each JSON object in the array according to ignore_index.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        field_path = curr_field
+        for elem in field_value:
+            if self.ignore_index:
+                new_field_path = field_path
+            else:
+                new_field_path = field_path + (field_value.index(elem), )
+
+            self.traverse(elem, new_field_path)
+
+    def process_type_field(self, field_value, curr_field):
+        """Check if a name field with type "type" has already appeared.
+
+        If had appeared with different type,raise an exception.
+        Otherwise, add object to a dict of types.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        if curr_field in self.fields_types_dict.keys():
+            if self.fields_types_dict[curr_field].field_type != field_value:
+                err_msg = "Type mismatch in schemas {0} and {1}."
+                old_schema = self.fields_types_dict[curr_field].schema_index
+                raise JSONSchemaCompatibilityException(
+                    old_schema,
+                    self.schema_name,
+                    err_msg.format(old_schema, self.schema_name)
+                )
+        else:
+            self.fields_types_dict[curr_field] = FieldToAdd(
+                schema_index=self.schema_name,
+                field_tuple=curr_field,
+                field_type=field_value
+            )
+
+    def check_special(self, curr_schema, curr_field):
+        """Check schema for allOf, anyOf and oneOf.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        keys = self.COLLECTION_KEYS.intersection(curr_schema)
+        for key in keys:
+            self.collect_special(
+                curr_schema[key],
+                curr_field
+            )
+
+    def collect_special(self, JSON_objects_list, curr_field):
+        """Operate a field with name 'allOf', 'anyOf' or 'oneOf'.
+
+        :param JSON_objects_list: List of JSON objects.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        for elem in JSON_objects_list:
+            self.traverse(elem, curr_field)
+
+    def traverse_dependencies(self, curr_schema, curr_field):
+        """Traverse and check dependencies in schema.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        """
+        properties = curr_schema['properties']
+        dependencies = curr_schema['dependencies']
+
+        while '$ref' in dependencies:
+            path = dependencies.get('$ref')
+            dependencies = self.resolver.resolve(path)[1]
+
+        self.traverse(dependencies, curr_field)
```

Comment:
The `dependencies` dict has not the same meaning and structure as the `properties` dict or the `root` dict. Better not call `traverse` on it.